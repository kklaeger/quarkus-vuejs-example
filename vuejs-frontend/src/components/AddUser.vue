<template>
  <div>
    <button class="button is-primary" @click="modalIsOpen = true">
      Add User
    </button>
    <div v-if="modalIsOpen">
      <b-modal active has-modal-card aria-role="dialog" aria-modal>
        <AddUserModal @close="modalIsOpen = false" />
      </b-modal>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import AddUserModal from "@/components/AddUserModal.vue";

@Component({
  components: { AddUserModal }
})
export default class AddUser extends Vue {
  private modalIsOpen = false;
}
</script>

<style scoped lang="scss"></style>
