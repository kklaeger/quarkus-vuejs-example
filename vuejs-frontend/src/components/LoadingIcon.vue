<template>
  <b-loading :is-full-page="false" v-model="isLoading"></b-loading>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
@Component({
  components: {}
})
export default class LoadingIcon extends Vue {
  private isLoading = true;
}
</script>

<style scoped lang="scss"></style>
