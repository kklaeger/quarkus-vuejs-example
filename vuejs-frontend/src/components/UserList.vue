<template>
  <div>
    <div v-if="this.users.length == 0">
      No users available :)
    </div>
    <div v-else>
      <ul>
        <UserListElement
          v-for="user in this.users"
          :key="user.id"
          :user="user"
        />
      </ul>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from "vue-property-decorator";
import UserListElement from "@/components/UserListElement.vue";
import { User } from "@/models/User";

@Component({
  components: { UserListElement }
})
export default class UserList extends Vue {
  @Prop({ required: true }) users!: User[];
}
</script>

<style scoped lang="scss"></style>
