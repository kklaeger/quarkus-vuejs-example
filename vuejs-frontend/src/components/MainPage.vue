<template>
  <section class="section">
    <div v-if="isLoading" class="loading">
      <LoadingIcon />
    </div>
    <div v-if="hasError" class="error">
      Unable to load users: {{ errorMsg }}
    </div>
    <div v-if="!isLoading && !hasError" class="container user-list">
      <AddUser class="addUser" />
      <UserList :users="users" />
    </div>
  </section>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import UserList from "@/components/UserList.vue";
import LoadingIcon from "@/components/LoadingIcon.vue";
import AddUser from "@/components/AddUser.vue";

@Component({
  components: { UserList, LoadingIcon, AddUser }
})
export default class MainPage extends Vue {
  mounted() {
    this.$store.dispatch("fetchUsers");
  }

  get users() {
    return this.$store.state.users;
  }

  get isLoading() {
    return this.$store.state.fetchUsersIsLoading;
  }

  get hasError() {
    return this.$store.state.fetchUsersHasError;
  }

  get errorMsg() {
    return this.$store.state.fetchUsersErrorMsg;
  }
}
</script>

<style scoped lang="scss">
@import "./../styles/variables.scss";
.loading {
  margin-bottom: 2rem;
}
.error {
  color: $error-message;
}
.addUser {
  margin-bottom: 3rem;
}
</style>
