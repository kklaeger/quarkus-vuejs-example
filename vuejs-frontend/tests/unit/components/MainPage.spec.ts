import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import MainPage from "@/components/MainPage.vue";
import UserList from "@/components/UserList.vue";
import LoadingIcon from "@/components/LoadingIcon.vue";
import AddUser from "@/components/AddUser.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

describe(MainPage.name + " component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render and match snapshot given isLoading=true.", () => {
    const state = {
      users: [],
      fetchUsersIsLoading: true,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false
    };
    const actions = {
      fetchUsers: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(MainPage, { store, localVue });

    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeTruthy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeTruthy();
    expect(wrapper.find(".user-list").exists()).toBeFalsy();
    expect(actions.fetchUsers).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render and match snapshot given isError=true.", () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: true,
      addUserIsLoading: false,
      addUserHasError: false,
      fetchUsersErrorMsg: "test"
    };
    const actions = {
      fetchUsers: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(MainPage, { store, localVue });
    expect(wrapper.find(".error").exists()).toBeTruthy();
    expect(wrapper.find(".error").text()).toEqual("Unable to load users: test");
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper.find(".user-list").exists()).toBeFalsy();
    expect(actions.fetchUsers).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render and match snapshot given isError=false and isLoading=false.", () => {
    const state = {
      users: [{ name: "user1" }],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false
    };
    const actions = {
      fetchUsers: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = shallowMount(MainPage, { store, localVue });
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper.find(".user-list").exists()).toBeTruthy();
    expect(wrapper.findAllComponents(UserList).exists()).toBeTruthy();
    expect(wrapper.findAllComponents(AddUser).exists()).toBeTruthy();
    expect(actions.fetchUsers).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });
});
