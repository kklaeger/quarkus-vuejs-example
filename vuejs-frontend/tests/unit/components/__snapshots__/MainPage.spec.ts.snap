// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component should render and match snapshot given isError=false and isLoading=false. 1`] = `
<section class="section">
  <!---->
  <!---->
  <div class="container user-list">
    <adduser-stub class="addUser"></adduser-stub>
    <userlist-stub users="[object Object]"></userlist-stub>
  </div>
</section>
`;

exports[`MainPage component should render and match snapshot given isError=true. 1`] = `
<section class="section">
  <!---->
  <div class="error">
    Unable to load users: test
  </div>
  <!---->
</section>
`;

exports[`MainPage component should render and match snapshot given isLoading=true. 1`] = `
<section class="section">
  <div class="loading">
    <loadingicon-stub></loadingicon-stub>
  </div>
  <!---->
  <!---->
</section>
`;
